tunnel/intra/doh/padding.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2019 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package doh
+
+import (
+	"golang.org/x/net/dns/dnsmessage"
+)
+
+const (
+	OptResourcePaddingCode = 12
+	PaddingBlockSize       = 128 // RFC8467 recommendation
+)
+
+const kOptRrHeaderLen int = 1 + // DOMAIN NAME
+	2 + // TYPE
+	2 + // CLASS
+	4 + // TTL
+	2 // RDLEN
+
+const kOptPaddingHeaderLen int = 2 + // OPTION-CODE
+	2 // OPTION-LENGTH
+
+// Compute the number of padding bytes needed, excluding
+// headers. Assumes that |msgLen| is the length of a raw DNS message
+// excluding any RFC7830 padding option, and that the message is fully
+// label-compressed.
+func computePaddingSize(msgLen int, hasOptRr bool, blockSize int) int {
+	// We'll always be adding a new padding header inside the OPT
+	// RR's data.
+	var extraPadding = kOptPaddingHeaderLen
+
+	// If we don't already have an OPT RR, we'll need to add its
+	// header as well!
+	if !hasOptRr {
+		extraPadding += kOptRrHeaderLen
+	}
+
+	var padSize int = blockSize - (msgLen+extraPadding)%blockSize
+	if padSize < 0 {
+		padSize *= -1
+	}
+	if padSize%blockSize == 0 {
+		padSize = 0
+	}
+	return padSize
+}
+
+func getPadding(msgLen int, hasOptRr bool) dnsmessage.Option {
+	optPadding := dnsmessage.Option{
+		Code: OptResourcePaddingCode,
+		Data: make([]byte, computePaddingSize(msgLen, hasOptRr, PaddingBlockSize)),
+	}
+	return optPadding
+}
+
+// Add EDNS padding, as defined in RFC7830, to a raw DNS message.
+func AddEdnsPadding(rawMsg []byte) ([]byte, error) {
+	var msg dnsmessage.Message
+	if err := msg.Unpack(rawMsg); err != nil {
+		return nil, err
+	}
+
+	// Search for OPT resource and save |optRes| pointer if possible.
+	var optRes *dnsmessage.OPTResource = nil
+	for _, additional := range msg.Additionals {
+		switch additional.Body.(type) {
+		case *dnsmessage.OPTResource:
+			optRes = additional.Body.(*dnsmessage.OPTResource)
+			break
+		}
+	}
+
+	if optRes != nil {
+		// If the message already contains padding, we will
+		// respect the application's padding.
+		return rawMsg, nil
+	}
+
+	// Build the padding option that we will need. We can't use
+	// the length of |rawMsg|, since its labels may be compressed
```

Comment:
Sounds tricky.  Can we add a test?  You could take the same `dnsmessage.Message`, pack it compressed and uncompressed, verify that the sizes are different, and then verify that the sizes are the same after padding.

I do feel I should point out that compression is unlikely to activate in queries, since it only applies when there are multiple names with a shared suffix, and queries normally contain only a single name.

Reply:
OK, so the library isn't flexible enough to allow me to choose compressed or not. I've sort of inverted your suggestion with two tests.

Test 1: Take bytes of a compressed query, unpack and repack with compression.  Check that the original and repacked queries have the same length.

Test 2: Take bytes of an *uncompressed* query, unpack and repack with compression. Check that the original query is larger than the repacked query.

These tests are really just verifying that the dnsmessage library does what we expect. We should also test that an uncompressed query comes out padded to the right size.

Reply:
> OK, so the library isn't flexible enough to allow me to choose compressed or not.

It looks like there might be a way:
https://godoc.org/golang.org/x/net/dns/dnsmessage#Builder.EnableCompression

> These tests are really just verifying that the dnsmessage library does what we expect.

That doesn't sound great.  I think the real goal is to test that this does the right thing when the input is not compressed (or is compressed differently from how `dnsmessage` would do it).

Reply:
You're right, it actually doesn't look too hard to use the `Builder` to control whether compression is enabled.

> I think the real goal is to test that this does the right thing when the input is not compressed (or is compressed differently from how dnsmessage would do it).

Absolutely.  To be clear, the "sanity check" tests on the behavior of the dnsmessage library are not the *only* tests.

I manually constructed the bytes of uncompressed and compressed queries, and wrote some tests using those.  I also have tests for queries with/without OPT and with/without padding.

Let me know if there's some situation I'm missing, and/or if you think I should explicitly use `Builder`.